compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala

```diff
@@ -329,7 +329,7 @@ trait TypeAssigner {
             errorType("ambiguous parent class qualifier", tree.pos)
         }
         val owntype =
-          if (mixinClass.exists) mixinClass.appliedRef
+          if (mixinClass.exists) mixinClass.typeRef
```

Comment:
I think neither old nor new version is right. What about this scenario?

    trait A[X]
    class B extends A[Int] {
      super[A] // should have type A[Int], not A, or A[X]
    }

Reply:
Yes, it's a bit weird, but if we change this then we need to change the pickling/unpickling scheme for super which currently does:
```scala
            case SUPER =>
              val qual = readTerm()
              val (mixId, mixTpe) = ifBefore(end)(readQualId(), (untpd.EmptyTypeIdent, NoType))
tpd.Super(qual, mixId, ctx.mode.is(Mode.InSuperCall), mixTpe.typeSymbol
```
So won't work with mix being an AppliedType.

Alternatively we could change SuperType#underyling to work like SuperType#superType, currently they're different:
```scala
  abstract case class SuperType(thistpe: Type, supertpe: Type) extends CachedProxyType with SingletonType {
    override def underlying(implicit ctx: Context): Type = supertpe
    override def superType(implicit ctx: Context): Type =
thistpe.baseType(supertpe.typeSymbol)
```


But note that in practice I haven't found a case where the current definition breaks something, e.g. in `findMember` we call `underlying` on `SuperType` but that ends up working out since `findMember` keeps around the prefix of the call as `pre` which contains all the information we need to figure out the type parameters.

Reply:
I believe `underlying` in supertype is fine.

I don't see right now why pickler/unpickler has to change. Should we not instead change the TypeAssigner to do the same as `SuperType$superType`?

Reply:
Coming back to this PR 2 years (!) later, I believe I'm still confused by why `underlying` and `superType` are different in `SuperType`, if I follow your advice of making `TypeAssigner` use the same logic as `superType` to set `superType`, then they would always be the same anyway. Would you mind taking another look at this?

Reply:
I don't have deep insights here either. `underlying` does not do `baseType` invocations, and `superType` should probably do that. 

That said, now think the change in TypeAssigner makes sense. If everything passes with that change, let's do it.